src/util/timemanager.cpp

```diff
@@ -59,133 +121,264 @@ void TimeManager::preSynchronization(double dt) {
-    ) != firstFutureKeyframe;
-
-    if (firstFutureKeyframe == keyframes.end()) {
-        // All keyframes are in the past.
-        // Consume the latest one.
-        const Keyframe<Time>& current = keyframes.back();
-        const Time& currentTime = current.data;
-        time().setTime(currentTime.j2000Seconds(), consumingTimeJump);
-        time().setDeltaTime(currentTime.deltaTime());
-        time().setPause(currentTime.paused());
-        _latestConsumedTimestamp = current.timestamp;
-    }
-    else {
-        const Keyframe<Time>& next = *firstFutureKeyframe;
-        const Time& nextTime = next.data;
-
-        if (firstFutureKeyframe != keyframes.begin()) {
-            const Keyframe<Time>& latest = *(firstFutureKeyframe - 1);
-            const Time& latestTime = latest.data;
-            // In case of unconsumed passed keyframes, let the last one
-            // determine whether the time should be paused or not.
-            // If there was a time jump or time is paused, apply it directly.
-            // Then consume the last keyframe.
-
-            time().setPause(latestTime.paused());
-            time().setTimeJumped(consumingTimeJump);
-            time().setDeltaTime(latestTime.deltaTime());
-
-            if (consumingTimeJump || latestTime.paused()) {
-                time().setTime(latestTime.j2000Seconds(), consumingTimeJump);
-            }
-            _latestConsumedTimestamp = latest.timestamp;
-        }
+    const bool hasFutureKeyframes = firstFutureKeyframe != keyframes.end();
+    const bool hasPastKeyframes = firstFutureKeyframe != keyframes.begin();
+
+    const auto lastPastKeyframe = hasPastKeyframes ?
+        (firstFutureKeyframe - 1) :
+        keyframes.end();
+
+    if (hasPastKeyframes && hasFutureKeyframes && !firstFutureKeyframe->data.jump) {
+        return interpolate(*lastPastKeyframe, *firstFutureKeyframe, applicationTime);
+    } else if (hasPastKeyframes) {
+        // Extrapolate based on last past keyframe
+        const double deltaApplicationTime = applicationTime - lastPastKeyframe->timestamp;
+        Time predictedTime = {
+            lastPastKeyframe->data.time.j2000Seconds() +
+            deltaApplicationTime *
+                (lastPastKeyframe->data.pause ? 0.0 : lastPastKeyframe->data.delta)
+        };
+        return TimeKeyframeData{
+            predictedTime,
+            lastPastKeyframe->data.delta,
+            false,
+            false
+        };
+    }
+    // As the last option, fall back on the current time.
+    return TimeKeyframeData{ _currentTime, _targetDeltaTime, _timePaused, false };
+}
 
-        // Do not interpolate with time jumping keyframes.
-        // Instead, wait until their timestamp and apply them directly.
-        if (nextTime.timeJumped()) {
-            time().advanceTime(dt);
-            return;
-        }
+void TimeManager::progressTime(double dt) {
+    static_cast<Time&>(_integrateFromTime) = _currentTime;
```

Comment:
That looks really ugly;  is there a better way of expressing it without doing a cast on the lvalue?